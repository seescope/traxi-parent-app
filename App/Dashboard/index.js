import React from 'react';
import lodash from 'lodash';
import { connect } from 'react-redux';
import { ScrollView, } from 'react-native';
import * as Animatable from 'react-native-animatable';
import Swiper from 'react-native-swiper';

import Spacing from '../Components/Spacing';
import { VERY_LIGHT_GREY } from '../Constants/Colours';
import KidCircle from './Components/KidCircle';
import Card from './Components/Card';
import TopApp from './Components/TopApp';
import TopCategory from './Components/TopCategory';
import PeakTime from './Components/PeakTime';
import RecentApp from './Components/RecentApp';

const containerStyle = {
  alignItems: 'center',
  paddingVertical: 36,
  paddingHorizontal: 8,
  backgroundColor: VERY_LIGHT_GREY,
};

const outerContainerStyle = {
  backgroundColor: VERY_LIGHT_GREY,
};

const getTodayUsage = peakTimes => ((peakTimes && peakTimes.week)
    ? (lodash.find(peakTimes.week, { name: 'Today'}) || { usage: 0}).usage
    : 0);

const DashboardScreen = ({ loading, kid, topApps, topCategories, peakTimes, recentApps }) =>
  <ScrollView style={outerContainerStyle} contentContainerStyle={containerStyle}>
    <Animatable.View
      useNativeDriver
      animation="bounceIn"
    >
      <KidCircle loading={loading} kid={kid} usage={getTodayUsage(peakTimes)} />
    </Animatable.View>

    <Spacing height={32} />

    <Card loading={loading} header="Top Apps" Component={TopApp} data={topApps} />

    <Card loading={loading} header="Top Categories" Component={TopCategory} data={topCategories} />

    <Card loading={loading} header="Peak Times" Component={PeakTime} data={peakTimes} />

    <Card loading={loading} header="Recent Apps" Component={RecentApp} data={recentApps} />
  </ScrollView>

DashboardScreen.propTypes = {
  topApps: React.PropTypes.object,
  topCategories: React.PropTypes.object,
  peakTimes: React.PropTypes.object,
  kid: React.PropTypes.object,
  recentApps: React.PropTypes.array,
  loading: React.PropTypes.bool.isRequired,
};

const mapStateToProps = ({ reports, kids, loading }) => ({
  kids,
  reports,
  loading,
});

const Dashboard = ({ reports, kids, loading }) =>
  <Swiper>
    {kids.map(kid => <DashboardScreen key={kid.UUID} loading={loading} kid={kid} {...reports[kid.UUID]} />)}
  </Swiper>

Dashboard.propTypes = {
  reports: React.PropTypes.object.isRequired,
  kids: React.PropTypes.array.isRequired,
  loading: React.PropTypes.bool.isRequired,
};

export default connect(mapStateToProps)(Dashboard);
