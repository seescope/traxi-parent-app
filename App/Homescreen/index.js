import React from 'react';
import { Dimensions, ScrollView, View, Text, Image } from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';

import Spacing from '../Components/Spacing';
import { GREY, VERY_LIGHT_GREY, LIGHT_BLUE, TRAXI_BLUE } from '../Constants/Colours';
import KidCircle from './Components/KidCircle';
import Card from './Components/Card';

const { width } = Dimensions.get('window');

const containerStyle = {
  alignItems: 'center',
  paddingVertical: 36,
  paddingHorizontal: 8,
  backgroundColor: VERY_LIGHT_GREY,
};

const outerContainerStyle = {
  backgroundColor: VERY_LIGHT_GREY,
};


const LOGOS = {
  facebook: 'http://is3.mzstatic.com/image/thumb/Purple111/v4/cc/73/53/cc735371-ee27-d48d-c8cc-51cf6c85290e/source/512x512bb.jpg',
  audible: 'http://is1.mzstatic.com/image/thumb/Purple122/v4/4f/98/93/4f989348-2928-302a-36b5-6df9ccf2592a/source/512x512bb.jpg',
};

const logoStyle = {
  borderRadius: 8,
  alignItems: 'center',
  justifyContent: 'center',
  height: 56,
  width: 56,
  marginRight: 24,
};

const rowStyle = {
  flexDirection: 'row',
  alignItems: 'center',
  marginBottom: 16,
  height: 56,
};

const innerHeaderTextStyle = {
  fontSize: 15,
  fontWeight: '400',
};

const innerSubheaderTextStyle = {
  fontSize: 15,
  fontWeight: '200',
};

const timeTextStyle = {
  width: 80,
};

const barStyle = {
  width: width - 32 - 96 - 16, 
  height: 16,
  backgroundColor: TRAXI_BLUE,
  borderRadius: 4,
};

const secondBarStyle = {
  width: (width - 32 - 96 - 16) / 2, 
  height: 16,
  backgroundColor: LIGHT_BLUE,
  borderRadius: 4,
};


export default () =>
  <ScrollView style={outerContainerStyle} contentContainerStyle={containerStyle}>
    <KidCircle />

    <Spacing height={32} />

    <Card header="Top Apps">

      <View style={rowStyle}>
        <Image source={{ uri: LOGOS.facebook }} style={logoStyle} />

        <View>
          <Text style={innerHeaderTextStyle}>Facebook</Text>
          <View style={barStyle} />
          <Text style={innerSubheaderTextStyle}>57 minutes</Text>
        </View>
      </View>

      <View style={rowStyle}>
        <Image source={{ uri: LOGOS.audible }} style={logoStyle} />

        <View>
          <Text style={innerHeaderTextStyle}>Audible</Text>
          <View style={secondBarStyle} />
          <Text style={innerSubheaderTextStyle}>23 minutes</Text>
        </View>
      </View>
    </Card>

    <Card header="Top Categories">
      <View style={rowStyle}>
        <Icon style={logoStyle} name="gamepad" size={52} color={GREY} />

        <View>
          <Text style={innerHeaderTextStyle}>Games</Text>
          <View style={barStyle} />
          <Text style={innerSubheaderTextStyle}>57 minutes</Text>
        </View>
      </View>

      <View style={rowStyle}>
        <Icon style={logoStyle} name="wrench" size={52} color={GREY} />

        <View>
          <Text style={innerHeaderTextStyle}>Tools</Text>
          <View style={secondBarStyle} />
          <Text style={innerSubheaderTextStyle}>23 minutes</Text>
        </View>
      </View>
    </Card>

    <Card header="Peak Times">
      <View style={rowStyle}>
        <View style={timeTextStyle}>
          <Text style={innerHeaderTextStyle}>9AM - 10AM</Text>
          <Text style={innerSubheaderTextStyle}>57 minutes</Text>
        </View>
        <View style={barStyle} />
      </View>

      <Spacing height={16} />

      <View style={rowStyle}>
        <View style={timeTextStyle}>
          <Text style={innerHeaderTextStyle}>6PM - 7PM</Text>
          <Text style={innerSubheaderTextStyle}>23 minutes</Text>
        </View>
        <View style={secondBarStyle} />
      </View>
    </Card>

    <Card header="Recent Apps">
      <View style={rowStyle}>
        <Image source={{ uri: LOGOS.facebook }} style={logoStyle} />
        <View>
          <Text style={innerHeaderTextStyle}>Facebook</Text>
          <Text style={innerSubheaderTextStyle}>15 minutes ago</Text>
        </View>
      </View>

      <View style={rowStyle}>
        <Image source={{ uri: LOGOS.audible }} style={logoStyle} />
        <View>
          <Text style={innerHeaderTextStyle}>Audible</Text>
          <Text style={innerSubheaderTextStyle}>23 minutes ago</Text>
        </View>
      </View>

    </Card>
  </ScrollView>
