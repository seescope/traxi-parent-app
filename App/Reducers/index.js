// @flow
/* eslint no-duplicate-imports: 0 */

import { combineReducers } from 'redux';
import parent from './Parent';
import kids from './Kids';
import setup from './Setup';
import reports from './Reports';

import type { ParentState } from './Parent';
import type { KidsState } from './Kids';
import type { SetupState } from './Setup';
import type { ReportsState } from './Reports';

export type RootState = {
  parentState: ParentState,
  kidsState: KidsState,
  setupState: SetupState,
  reportsState: ReportsState
};

const rootState = {
  parentState: parent,
  kidsState: kids,
  setupState: setup,
  reportsState: reports,
};

export default combineReducers(rootState);
