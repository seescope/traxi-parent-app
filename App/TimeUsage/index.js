/* eslint global-require: "off" */

import React, { PropTypes } from 'react';
import { Image, StyleSheet, Dimensions } from 'react-native';
import Header from './Header';
import UsageItems from './UsageItems';
import { connect } from 'react-redux';

const { width, height } = Dimensions.get('window');

const STYLE = StyleSheet.create({
  container: {
    flex: 1,
    width,
    height,
    alignItems: 'flex-start',
  },
});


export const TimeUsageComponent = ({ kid, date, report }) => (
  <Image source={require('../Images/background.jpg')} style={STYLE.container}>
    <Header date={date} kid={kid} />
    <UsageItems report={report} />
  </Image>
);

TimeUsageComponent.propTypes = {
  kid: PropTypes.object.isRequired,
  date: PropTypes.string.isRequired,
  report: PropTypes.array.isRequired,
};

export default connect((state, ownProps) => (
  {
    kid: state.selectedKid,
    date: ownProps.date,
    report: ownProps.report,
  }
))(TimeUsageComponent);
