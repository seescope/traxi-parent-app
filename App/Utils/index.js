/* eslint no-unused-vars: ["error", {"args": "none"}] */

import moment from 'moment';
import { Platform } from 'react-native';
import { Crashlytics } from 'react-native-fabric';
import Analytics from 'react-native-analytics';

export const firstName = name => name && name.split(' ')[0];

export const relativeDate = inputDate => {
  const date = moment(inputDate);

  if (date.isSame(moment(), 'day')) {
    return 'Today';
  }

  const lastNight = moment().subtract(1, 'days');

  if (date.isSame(lastNight, 'day')) {
    return 'Yesterday';
  }


  return date.format('dddd');
};

export const timeRange = inputHour => {
  const hour = moment(inputHour, 'hhA');
  const nextHour = hour.clone();
  nextHour.add(1, 'hour');

  return `${hour.format('hA')} to ${nextHour.format('hA')}`;
};

export const isIOS = (Platform.OS === 'ios');

export const listOfNumbers = length => {
  const list = [];
  for (let i = 0; i < length; i++) {
    list.push(i);
  }

  return list;
};

const niceNames = {
  'Maps - Navigation & Transit': 'Google Maps',
  'Chrome Browser - Google': 'Chrome',
};

export const getAppNiceName = name => niceNames[name] || name;

export const isToday = date => moment(date).isSame(moment(), 'day');

export const logError = error => {
  console.error(error);
  return (isIOS ?
    Crashlytics.recordError(error.toString()) :
    Crashlytics.logException(error.toString()));
};

export const loggingMiddleware = store => next => action => {
  Crashlytics.log(JSON.stringify(action));
  Crashlytics.log(JSON.stringify(store));

  return next(action);
};

export const trackingMiddleware = store => next => action => {
  // If this isn't a screen change, we're not interested.
  if (action.type !== 'REACT_NATIVE_ROUTER_FLUX_PUSH') {
    return next(action);
  }

  // Grab the name of the screen from the flux action.
  const { key } = action;
  Analytics.screen(key);

  return next(action);
};

export const experimentViewed = (variantName) => {
  Analytics.identify({
    price: variantName,
  });
};
