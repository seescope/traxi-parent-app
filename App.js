import React from 'react';
import { AsyncStorage } from 'react-native';
import * as Firebase from 'firebase';
import I18n from 'react-native-i18n';
import codePush from 'react-native-code-push';

import ParentApp from './App/Containers/ParentApp';
import Translation from './App/Constants/Translation';
import Analytics from 'react-native-analytics';
import OneSignal from 'react-native-onesignal';
import { compose, createStore, applyMiddleware } from 'redux';
import { persistStore, autoRehydrate } from 'redux-persist';
import ReduxThunk from 'redux-thunk';
import RootReducer from './App/Reducers';
import { backButtonHandler, loggingMiddleware } from './App/Utils';
import bootApp from './App/AsyncActions/BootApp';
import trackingMiddleware from './App/Middleware/Tracking';
import firebaseTrackingMiddleware from './App/Middleware/FirebaseTracking';
import impersonateParent from './App/AsyncActions/ImpersonateParent'; // eslint-disable-line
// import actions from './parentAppActions.json';
// import replayActions from './App/Middleware/ReplayActions';

I18n.fallbacks = true;
I18n.translations = Translation;

const config = {
  apiKey: 'AIzaSyDEq9qfwendZJ6yiyDgtjGCjWSS9PSYWLU',
  authDomain: 'traxiapp.firebaseapp.com',
  databaseURL: 'https://traxiapp.firebaseio.com',
  projectId: 'project-946779331638130823',
  storageBucket: 'project-946779331638130823.appspot.com',
  messagingSenderId: '204102393429',
};

try {
  Firebase.initializeApp(config);
} catch (e) {
  console.warn('Error initialising Firebase', e);
}

export default class extends React.Component {
  constructor(props) {
    super(props);
    this.store = createStore(
      RootReducer,
      undefined,
      compose(
        applyMiddleware(
          ReduxThunk,
          loggingMiddleware,
          trackingMiddleware,
          firebaseTrackingMiddleware
        ),
        autoRehydrate()
      )
    );

    // Boot normally. Comment this out if testing.
    persistStore(this.store, { storage: AsyncStorage }, () =>
      this.store.dispatch(bootApp()));

    // TEST_PARENT_UUID, useful for... testing.

    // const TEST_PARENT_UUID = 'ee716cfb-76dc-4e86-9e43-4577a378f255';

    // Uncomment to replay actions.
    // setTimeout(
    //   () =>
    //     replayActions(
    //       TEST_PARENT_UUID,
    //       actions,
    //       this.store.dispatch
    //     ),
    //   1000
    // );

    // Uncomment to impersonate a parent for testing.
    // this.store.dispatch(impersonateParent(TEST_PARENT_UUID));
    // console.warn('Impersonating parent', TEST_PARENT_UUID);
  }

  componentWillMount() {
    OneSignal.addEventListener('received', this.onReceived);
    OneSignal.addEventListener('opened', this.onOpened);
  }

  componentWillUnmount() {
    OneSignal.removeEventListener('received', this.onReceived);
    OneSignal.removeEventListener('opened', this.onOpened);
  }

  onReceived() {
    Analytics.track('Notification received');
  }

  onOpened() {
    Analytics.track('Notification opened');
  }

  async codePushStatusDidChange(status) {
    switch (status) {
      case codePush.SyncStatus.UPDATE_INSTALLED: {
        const updateMetadata = await codePush.getUpdateMetadata();
        if (updateMetadata) {
          Analytics.track('CodePush update installed', {
            label: updateMetadata.label,
          });
        }
        break;
      }
      default: {
        const updateMetadata = await codePush.getUpdateMetadata();
        console.log('CodePush status changed', updateMetadata); // eslint-disable-line
      }
    }
  }

  render() {
    return (
      <ParentApp
        store={this.store}
        backButtonHandler={() => backButtonHandler(this.store)}
      />
    );
  }
}
