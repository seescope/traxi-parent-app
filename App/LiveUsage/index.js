import React, { PropTypes } from 'react';
import { ActivityIndicator } from 'react-native';
import * as Animatable from 'react-native-animatable';
import { TRAXI_ORANGE, NEUTRAL } from '../Constants/Colours';
import Firebase from 'firebase';
import Background from '../Components/Background';
import { firstName, isIOS } from '../Utils';

const liveUsage = {
  paddingTop: 64,
  alignItems: 'center',
  flex: 1,
  backgroundColor: 'white',
};


const headerText = {
  color: NEUTRAL,
  textAlign: 'center',
  fontWeight: '300',
  fontSize: isIOS ? 42 : 18,
  marginBottom: 8,
  backgroundColor: 'transparent',
};

const activityIndicatorStyle = {
  marginBottom: 32,
};

class LiveUsage extends React.Component {
  constructor(props) {
    super(props);
    const { UUID } = props.kid;
    this.state = {
      domains: [],
    };
    this.firebase = new Firebase(`https://traxiapp.firebaseio.com/kids/${UUID}/live`);
  }

  componentDidMount() {
    this.firebase.on('child_added', data => {
      const newDomain = data.val().d;
      this.setState({ domains: [newDomain, ...this.state.domains] });
    });
  }

  render() {
    const kidName = firstName(this.props.kid.name);
    return (
      <Background style={liveUsage}>
        <ActivityIndicator size="large" color={TRAXI_ORANGE} style={activityIndicatorStyle} />
        {this.state.domains.length === 0 &&
          <Animatable.Text style={headerText} animation="fadeInDown">
            Connecting to {kidName}'s device..
          </Animatable.Text>}
        {this.state.domains.map((d, i) =>
          <Animatable.Text key={i} style={headerText} animation="fadeInDown">{d}</Animatable.Text>)
        }
      </Background>
    );
  }
}

LiveUsage.propTypes = {
  kid: PropTypes.object.isRequired,
};

export default LiveUsage;
