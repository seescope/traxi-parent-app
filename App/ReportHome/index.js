import React, { PropTypes } from 'react';
import {
  ActivityIndicator,
  ScrollView,
  View,
  StyleSheet,
  Dimensions,
} from 'react-native';
import { connect } from 'react-redux';
import { Actions } from 'react-native-router-flux';
import Analytics from 'react-native-analytics';
import moment from 'moment';

import Circle from './Circle';
import KidTile from './KidTile';
import {
  WHITE,
} from '../Constants/Colours';
import { firstName } from '../Utils';
import Button from '../Components/Button';
import HeaderText from '../Components/HeaderText';
import Background from '../Components/Background';
import { selectKid as selectKidAction } from '../Actions/Actions';

const { width, height } = Dimensions.get('window');

const style = StyleSheet.create({
  outerContainer: {
    width,
  },
  bottomContainer: {
    alignItems: 'center',
    width,
    flex: 1,
  },
  container: {
    minHeight: height,
    alignItems: 'center',
  },
  buttonContainer: {
    height: width / 2,
    justifyContent: 'center',
  },
});


export const buildProps = (kid, reports) => {
  let report;
  let loading;

  const kidReport = reports[kid.UUID];

  if (!kidReport) {
    loading = true;
    report = {};
  } else {
    // Real values.
    loading = false;
    report = reports[kid.UUID];
  }

  return {
    loading,
    report,
    kid,
  };
};

const getCircle = (kid, reports) => {
  const today = moment().format('YYYY-MM-DD');
  return reports[kid.UUID]
    && reports[kid.UUID][today]
    && reports[kid.UUID][today].circles[1];
};

const ReportHome = ({ kids, reports, loading, selectKid }) => (
  <ScrollView contentContainerStyle={style.container} style={style.outerContainer}>
    {kids.map(kid => <KidTile key={kid.UUID} kid={kid} onPress={() => selectKid(kid)}>
      <HeaderText>{firstName(kid.name)}</HeaderText>
      {getCircle(kid, reports) && <Circle {...getCircle(kid, reports)} />}
      {!getCircle(kid, reports) && loading && <ActivityIndicator size="large" color={WHITE} />}
    </KidTile>)}

    <Background style={style.bottomContainer}>
      <View style={style.buttonContainer}>
        <Button primary={false} onPress={() => Actions.walkthrough()}>Add another child</Button>
      </View>
    </Background>
  </ScrollView>
);

ReportHome.propTypes = {
  reports: PropTypes.object.isRequired,
  kids: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  selectKid: PropTypes.func.isRequired,
};

export const mapStateToProps = state => ({
  loading: state.loading,
  reports: state.reports,
  kids: state.profile.kids,
});

export const mapDispatchToProps = dispatch => ({
  selectKid: (kid) => {
    dispatch(selectKidAction(kid));
    Analytics.track('Viewed Report', {
      UUID: kid.UUID,
    });
    Actions.weekView();
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ReportHome);
