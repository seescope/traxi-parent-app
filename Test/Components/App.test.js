import React from 'react';
import App from '../../App';
import renderer from 'react-test-renderer';

jest.mock('../../App/AsyncActions/ImpersonateParent', () =>
  jest.fn(() => ({
    type: 'TEST_IMPERSONATE',
  })));
jest.mock('../../App/AsyncActions/BootApp', () =>
  jest.fn(() => ({
    type: 'TEST_BOOT_APP',
  })));

import bootApp from '../../App/AsyncActions/BootApp';
import impersonateParent from '../../App/AsyncActions/ImpersonateParent';

it('renders the root compontent correctly', () => {
  const tree = renderer.create(<App />).toJSON();

  expect(tree).toMatchSnapshot();
  expect(bootApp).toHaveBeenCalled();
  expect(impersonateParent).not.toHaveBeenCalled();
});
