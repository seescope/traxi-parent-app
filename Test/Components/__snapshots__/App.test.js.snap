// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the root compontent correctly 1`] = `
<View>
  <View
    component={[Function]}
    initial={true}
  />
  <View
    component={[Function]}
  />
  <View
    component={[Function]}
  />
  <View
    component={[Function]}
  />
  <View
    component={[Function]}
  />
  <View
    component={[Function]}
  />
  <View
    component={[Function]}
  />
  <View
    component={[Function]}
  />
  <View
    component={[Function]}
    initial={false}
  />
  <View
    component={[Function]}
  />
  <View
    component={[Function]}
  />
  <View
    component={[Function]}
  />
</View>
`;
